client.cfg

```
@@ -51,21 +51,15 @@ description: Cancel the actual running print
 rename_existing: CANCEL_PRINT_BASE
 gcode:
   ##### get user parameters or use default #####
-  {% set macro_found = True if printer['gcode_macro _CLIENT_VARIABLE'] is defined else False %}
-  {% set client = printer['gcode_macro _CLIENT_VARIABLE'] %}
-  {% set allow_park = False if not macro_found
-                 else False if client.park_at_cancel is not defined
-                 else True  if client.park_at_cancel|lower == 'true'
-                 else False %}
-  {% set retract = 5.0  if not macro_found else client.cancel_retract|default(5.0)|abs %}
+  {% set client = printer['gcode_macro _CLIENT_VARIABLE'] | default({}) %}
+  {% set allow_park = client.park_at_cancel | default(false) | lower == 'true' %}
+  {% set retract = client.cancel_retract | default(5.0) | abs %}
   ##### define park position #####
-  {% set park_x = ""                                    if not macro_found
-             else ""                                    if client.park_at_cancel_x is not defined
-             else "X=" + client.park_at_cancel_x|string if client.park_at_cancel_x is not none %}
-  {% set park_y = ""                                    if not macro_found
-             else ""                                    if client.park_at_cancel_y is not defined
-             else "Y=" + client.park_at_cancel_y|string if client.park_at_cancel_y is not none %}
-  {% set custom_park = True if (park_x|length > 0 or park_y|length > 0) else False %}
+  {% set park_x = "" if (client.park_at_cancel_x | default(none) is none)
+            else "X=" ~ client.park_at_cancel_x %}
+  {% set park_y = "" if (client.park_at_cancel_y | default(none) is none)
+            else "Y=" ~ client.park_at_cancel_y %}
```

Unless I am missing something, this can be further simplified:

```suggestion
  {% set park_x = "" if (client.park_at_cancel_x is not defined)
            else "X=" ~ client.park_at_cancel_x %}
  {% set park_y = "" if (client.park_at_cancel_y is not defined)
            else "Y=" ~ client.park_at_cancel_y %}
```

---

That has a different result in the case where `park_at_cancel_x` is explicitly set to `None`. `park_x` should end up as the empty string in that case, but if we just check whether it is defined it will be `X=None`

```
$ cat test.j2
{% set client = {'park_at_cancel_x': none} %}
{% set park_x = "" if (client.park_at_cancel_x | default(none) is none)
    else "X=" ~ client.park_at_cancel_x %}
{% set park_x_def = "" if (client.park_at_cancel_x is not defined)
    else "X=" ~ client.park_at_cancel_x %}
Original logic: "{ park_x }"
Just checking for defined: "{ park_x_def }"
```
results in
```
$ ./template.py test.j2



Original logic: ""
Just checking for defined: "X=None"
```

---

Fair, though I should point out that the original code **does** allow for `none`, thus it would match the output of "X=None" as on this approach.

(side note, I normally use this site to test Jinja2: http://jinja.quantprogramming.com/)

---

Well, that's where we get into the original structure being confusing and hard to read. It allows for `none`, but it has the same effective result as my version:
```
  {% set park_x = ""                                    if not macro_found
             else ""                                    if client.park_at_cancel_x is not defined
             else "X=" + client.park_at_cancel_x|string if client.park_at_cancel_x is not none %}
```
Jinja's inline `if` has an [implicit else](https://jinja.palletsprojects.com/en/3.1.x/templates/#if-expression), so the expression `"X=" + client.park_at_cancel_x|string if client.park_at_cancel_x is not none` returns a [jinja2.Undefined](https://jinja.palletsprojects.com/en/3.1.x/api/#jinja2.Undefined) object when `client.park_at_cancel_x` is `none`. When that object is treated as a string it returns an empty string.